Add property EJBCA_AJP_PORT_NUMBER for AJP port, EJBCA_SMTP_* for email and EJBCA_INITSCRIPTS_DIR for custom startup scripts

BlueIcarus opened the pull request:
**Description of the change**

1. Added property to enable the AJP protocol on the Wildfly container. The AJP protocol can be used in combination with a frontend Apache reverse proxy. The AJP protocol automatically adds details of the client certificate to forwarded data. 
2. Added properties to configure SMTP settings on the Wildfly container. This allows EJBCA to send out email notifications
3. Added property to add custom INIT scripts

**Benefits**

1. This allows one to setup a Apache reverse proxy with client certificate authentication and forward the client certificate data to the Wildfly container for further validation. This can increase security do to the fact that the Wildfly instance is not directly connected to the internet (or untrusted network).
2. This allows one to setup SMTP email notification for usage within EJBCA. For example to send email notification on certificates that will expire (and other email notification types).
3. This allows one to add custom INIT script that do other things on the system. For example allows one to configure outgoing http proxy settings on Wildfly.

**Possible drawbacks**

1. AJP is a Apache protocol mainly used by Wildfly, Tomcat and Apache and other reverse proxies might not support this protocol.
2. The current configuration might only work with the settings specified here. There might be SMTP related settings that currently cant be configured.
3. This might cause unknown startup issues if used incorrectly.

**Applicable issues**

1. Without this setup you can not use a reverse Apache proxy with client certificate authentication.
2. Without this setup you can't use the email notifications that can be used from within EJBCA.
3. Without this setup you will need to make manual changes to the container (or create a new one) to for example set proxy settings in Wildfly.

Mauraza commented:
Hi @BlueIcarus

Thanks for the PR!! It looks good to me. We are going to do some testing before merging.

Mauraza commented:
I'm testing the changes but appears some errors, I will continue investigating.

BlueIcarus commented:
I had the same challenge. I have updated a couple of things on an internal GIT repository. I will merge those changes in here as soon as I have tested them locally.

BlueIcarus commented:
As I mentioned before I updated the configuration some more. These commits should solve the problem with the AJP port. I was working on this on my own internal GIT and did not really do a proper job on naming the pushes. You should merge these changes into a single commit when merging it into a branch.

These commits not only fix the AJP problem (as I mentioned). They also allow you to configure the SMTP settings so you can use the email functionality. Finally I also added an option to include custom scripts. That is something we always do when running Docker containers that contain Wildfly. We use this (for example) to allow users to add custom startup instructions. Things like outgoing proxies can be configured to Wildfly so it can connect to the internet when using private proxies. And there are countless other things you might be able to solve using this feature. Hope this helps and you can merge it into a release version!

javsalgar commented on 6/debian-10/rootfs/opt/bitnami/scripts/libejbca.sh:
```diff
@@ -237,6 +249,16 @@ ejbca_stop_wildfly() {
 
+ejbca_custom_scripts() {
+    info "Running custom scripts..."
+    FILES=/bitnami/custom-scripts/*
```
In other applications we have an env var called `*_INITSCRIPTS_DIR`, so I set `EJBCA_INITSCRIPTS_DIR`

BlueIcarus commented:
I updated the title of this pull request and also updated the initial text related to this pull request. This will make it more clear to everyone what these changes imply.

Mauraza commented:
We are doing some internal tests, I will update this thread when we have more information.

BlueIcarus commented:
Any progress on those internal tests? I'm currently using my own personal Docker container with these changes but I would prefer to just use the one from Docker Hub so it holds all the latest changes (version updates of EJBCA).

Mauraza commented:
Hi @BlueIcarus,

I'm getting this error in the test
```
ejbca_1    | ejbca 11:09:06.05 INFO  ==> Configure email
ejbca_1    | {"outcome" => "success"}
ejbca_1    | 11:09:09,257 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-8) WFLYMAIL0002: Unbound mail session [java:/EjbcaMail]
ejbca_1    | 11:09:09,257 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-8) WFLYMAIL0001: Bound mail session [java:/EjbcaMail]
ejbca_1    | {"outcome" => "success"}
ejbca_1    | The closing '"' is missing.
```

Did you get that error?

BlueIcarus commented:
I found the cause of that issue. I was testing it with email setting but did not test it without email settings. There was one to many " in one of the config updates.

PaleoCrafter commented:
What is the current state of this PR? I'm trying to run EJBCA behind an SSL-terminating reverse proxy right now and have to jump through a few hoops to make it accept the passed-through client certificate.

BlueIcarus commented:
Well for me this works as it should. I have a private copy of this repository where I merge any changes from GitHub into so I have a up-to-date version of this container. Why its not merged, I cant tell you.

javsalgar commented:
It seems there are merge issues. Could you check them?

javsalgar commented:
@Mauraza is on PTO. As soon as she comes back she will provide more feedback on the PR.

Mauraza commented:
Hi @BlueIcarus,

I'm going to test the changes, I will update the thread.

javsalgar commented on 7/debian-10/rootfs/opt/bitnami/scripts/libejbca.sh:
```diff
@@ -237,6 +249,18 @@ ejbca_stop_wildfly() {
 
+ejbca_custom_scripts() {
+    info "Running custom scripts..."
+    if [[ -d "/bitnami/custom-scripts/" ]]; then
```
Could you change the path to `/docker-entrypoint-init.d` .

Use this code as an example:

https://github.com/bitnami/bitnami-docker-wordpress/blob/master/5/debian-10/rootfs/post-init.sh

javsalgar commented:
It seems there are merge issues, could you fix them?

BlueIcarus reviewed (commented):
This should do the trick

javsalgar commented:
I pinged @Mauraza as she is working on the newer major version and I'm not sure if this is still compatible.
